test/core/common/scale_test.cpp

```diff
@@ -0,0 +1,363 @@
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 1);
+  auto value = header.second;
+  ASSERT_EQ(value, 64);  // MAX_UI8 + 1
+}
+
+TEST(Scale, compactDecodeMaxUi16) {
+  // decode MAX_UI16 == 2^14 - 1 == 16383
+  std::vector<uint8_t> bytes = {
+      0b11111101,  // 253
+      0b11111111   // 255
+  };
+
+  compact::Decoder decoder{bytes};
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 1);
+  auto value = header.second;
+  ASSERT_EQ(value, 16383);  // MAX_UI16
+}
+
+TEST(Scale, compactDecodeMinUi32) {
+  // decode MIN_UI32 := MAX_UI16 + 1 = 2^14 = 16384
+  std::vector<uint8_t> bytes = {
+      0b00000010,  // 2
+      0b00000000,  // 0
+      0b00000001,  // 1
+      0b00000000   // 0
+  };
+
+  compact::Decoder decoder{bytes};
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 2);
+  auto value = header.second;
+  ASSERT_EQ(value, 16384);  // MAX_UI16+1
+}
+
+TEST(Scale, compactDecodeMaxUi32) {
+  // decode MAX_UI32 := 2^30 - 1 == 1073741823
+  std::vector<uint8_t> bytes = {
+      0b11111110,  // 254
+      0b11111111,  // 255
+      0b11111111,  // 255
+      0b11111111   // 255
+  };
+
+  compact::Decoder decoder{bytes};
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 2);
+  auto value = header.second;
+  ASSERT_EQ(value, 1073741823);  // MAX_UI32
+}
+
+TEST(Scale, compactEncodeUint8) {
+  // encode MAX_UI8 := 63
+  compact::encodeUint8(63).match(
+      [&](const expected::Value<ByteArray> &res) {
+        ASSERT_EQ(res.value, std::vector<uint8_t>({252}));
+      },
+      [&](const expected::Error<compact::EncodeBigIntegerError> &) { FAIL(); });
+
+  // encode MIN_UI16 := MAX_UI8 + 1
+  compact::encodeUint8(64).match(
+      [&](const expected::Value<ByteArray> &res) {
+        ASSERT_EQ(res.value, std::vector<uint8_t>({1, 1}));
+      },
+      [&](const expected::Error<compact::EncodeBigIntegerError> &) { FAIL(); });
+
+  // encode some UI16
+  compact::encodeUint8(255).match(
+      [&](const expected::Value<ByteArray> &res) {
+        ASSERT_EQ(res.value, std::vector<uint8_t>({253, 3}));
+      },
+      [&](const expected::Error<compact::EncodeBigIntegerError> &) { FAIL(); });
+
+  // encode some other UI16
+  compact::encodeUint16(511).match(
+      [&](const expected::Value<ByteArray> &res) {
+        ASSERT_EQ(res.value, std::vector<uint8_t>({253, 7}));
+      },
+      [&](const expected::Error<compact::EncodeBigIntegerError> &) { FAIL(); });
+}
+
+TEST(Scale, compactEncodeUint16) {
```

Comment:
Same, please split that into several test cases

Reply:
Refactored tests and methods.